### PR TITLE
FEAT-#6918: Add auto mode to the lazy execution.

### LINKED ISSUE #6918: FEAT: Add 'auto' mode to the lazy execution.
The `MODIN_LAZY_EXECUTION` env var should have 3 options:
`Auto` - the execution mode is chosen by the engine for each operation (default value).
`On`   - the lazy execution is performed wherever it's possible.
`Off`  - the lazy execution is disabled.

### PR DESCRIPTION
**Pull request by AndreyPavlenko**
## What do these changes do?

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #6918 
- [x] tests added and passing
- [ ] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/config/envvars.py [hunk 1]
```diff
@@ -823,10 +823,18 @@ class ReadSqlEngine(EnvironmentVariable, type=str):
...
+
+    Supported values:
+        `Auto` - the execution mode is chosen by the backend for each operation (default value).
+        `On`   - the lazy execution is performed wherever it's possible.
+        `Off`  - the lazy execution is disabled.
+    """
 
     varname = "MODIN_LAZY_EXECUTION"
+    choices = ("Auto", "On", "Off")
     default = False
```
Comment (dchigarev) [inline modin/config/envvars.py:838]: ?
```suggestion
    default = "Off"
```

File: modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition.py [hunk 2]
```diff
@@ -377,3 +375,53 @@ def _get_index_and_columns(df):  # pragma: no cover
...
+            self._eager_exec_func(func, *args, **kwargs)
+
+        PandasOnRayDataframePartition.apply = (
+            PandasOnRayDataframePartition._eager_exec_func
+        )
+        PandasOnRayDataframePartition.add_to_apply_calls = eager_exec
+
+
+_set_lazy_exec_mode(LazyExecution.get(), None)
+LazyExecution.put = _set_lazy_exec_mode
```
Comment (dchigarev) [inline modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition.py:427]: I would better use the [subscribe interface](https://github.com/modin-project/modin/blob/5f6f80942c53012a3e0e85c0479b3cc4f56253cf/modin/config/pubsub.py#L283)
```suggestion
LazyExecution.subscribe(_set_lazy_exec_mode)
```

File: modin/config/envvars.py [hunk 3]
```diff
@@ -816,11 +816,19 @@ class ReadSqlEngine(EnvironmentVariable, type=str):
...
 
 
-class LazyExecution(EnvironmentVariable, type=bool):
-    """Prefer the lazy execution, when it's possible."""
+class LazyExecution(EnvironmentVariable, type=str):
+    """
+    Lazy execution mode.
+
+    Supported values:
+        `Auto` - the execution mode is chosen by the backend for each operation (default value).
```
Comment (YarShev) [inline modin/config/envvars.py:824]:
```suggestion
        `Auto` - the execution mode is chosen by the engine for each operation (default value).
```

File: modin/test/storage_formats/pandas/test_internals.py [hunk 4]
```diff
@@ -2421,3 +2421,50 @@ def test_groupby_index_dtype(self):
...
+@pytest.mark.parametrize("mode", [None, "Auto", "On", "Off"])
+def test_ray_lazy_exec_mode(mode):
+    import ray
+
+    from modin.config import LazyExecution
+    from modin.core.execution.ray.common.deferred_execution import DeferredExecution
+    from modin.core.execution.ray.common.utils import ObjectIDType
+    from modin.core.execution.ray.implementations.pandas_on_ray.partitioning import (
+        PandasOnRayDataframePartition,
+    )
```
Comment (YarShev) [inline modin/test/storage_formats/pandas/test_internals.py:2436]: This should be at the top of this file. Why here?

File: modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition.py [hunk 5]
```diff
@@ -377,3 +375,45 @@ def _get_index_and_columns(df):  # pragma: no cover
...
+    if mode == "Auto":
+        PandasOnRayDataframePartition.apply = (
+            PandasOnRayDataframePartition._eager_exec_func
+        )
+        PandasOnRayDataframePartition.add_to_apply_calls = (
+            PandasOnRayDataframePartition._lazy_exec_func
+        )
+    elif mode == "On":
+
+        def lazy_exec(self, func, *args, **kwargs):
```
Comment (YarShev) [inline modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition.py:399]: Why do we need this wrapper?

File: modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition.py [hunk 6]
```diff
@@ -377,3 +375,45 @@ def _get_index_and_columns(df):  # pragma: no cover
...
+            return self._lazy_exec_func(func, *args, length=None, width=None, **kwargs)
+
+        PandasOnRayDataframePartition.apply = lazy_exec
+        PandasOnRayDataframePartition.add_to_apply_calls = (
+            PandasOnRayDataframePartition._lazy_exec_func
+        )
+    elif mode == "Off":
+
+        def eager_exec(self, func, *args, length=None, width=None, **kwargs):
```
Comment (YarShev) [inline modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition.py:408]: same question

File: modin/test/storage_formats/pandas/test_internals.py (hunk 4 above)
Comment (AndreyPavlenko) [reply modin/test/storage_formats/pandas/test_internals.py:2436]: This test is ray specific and is not executed for other engines.

File: modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition.py (hunk 5 above)
Comment (AndreyPavlenko) [reply modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition.py:399]: apply() and add_to_apply_calls() functions have different set of parameters.

Comment (YarShev) [reply modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition.py:399]: I see, we should align the parameters for both methods at some point.

File: modin/test/storage_formats/pandas/test_internals.py (hunk 4 above)
Comment (YarShev) [reply modin/test/storage_formats/pandas/test_internals.py:2436]: I see

**Review by YarShev — APPROVED** [review]

**Review by dchigarev — APPROVED** [review]

File: modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition.py [hunk 7]
```diff
@@ -125,11 +125,9 @@ def apply(self, func: Callable, *args, **kwargs):
         If ``LazyExecution`` is enabled, the function is not applied immediately,
         but is added to the execution tree.
```
Comment (anmyachev) [inline modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition.py:126]: @AndreyPavlenko should this also be removed?

Comment (AndreyPavlenko) [reply modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition.py:126]: Thanks. To be removed in a separate PR.

**Pull request merged**